src/librustdoc/clean/utils.rs

```diff
@@ -414,7 +414,10 @@ crate fn resolve_type(cx: &mut DocContext<'_>, path: Path, id: hir::HirId) -> Ty
             return Generic(kw::SelfUpper);
         }
         Res::Def(DefKind::TyParam, _) if path.segments.len() == 1 => {
-            return Generic(Symbol::intern(&format!("{:#}", path.print(&cx.cache))));
+            // FIXME(camelid): I hope passing 0 as the depth doesn't break anything....
+            //                 Then again, I think used to be 0 anyway through CURRENT_DEPTH
+            //                 because (I think) rendering hadn't started yet.
+            return Generic(Symbol::intern(&format!("{:#}", path.print(&cx.cache, 0))));
```

Comment:
I think same as above.

Reply:
This change makes me nervous, can you say why you think `rendering` hadn't started?

I don't know exactly when this path is hit, `path.segment.len() == 1` seems really strange because it doesn't leave any space for the crate name and I don't see any other reason to special case 1 path segment. Maybe add an unconditional `panic!()` so we have a test case?

Reply:
> This change makes me nervous

I understand, these changes make me a bit nervous too :)

> can you say why you think rendering hadn't started?

My understanding is that all `clean`ing happens before rendering, so `CURRENT_DEPTH` is always its initial value of `0`. Is that not true?

> Maybe add an unconditional `panic!()` so we have a test case?

Hmm, can you explain more what you mean?

Reply:
> My understanding is that all cleaning happens before rendering, so CURRENT_DEPTH is always its initial value of 0. Is that not true?

Ok, this makes sense to me. The depth is only updated in `context.rs`:

```

html/render/context.rs
22:    CURRENT_DEPTH, INITIAL_IDS,
107:        CURRENT_DEPTH.with(|slot| {
421:        CURRENT_DEPTH.with(|s| s.set(0));
```
and cleaning never happens in `html`/:
```
$ rg '\.clean' html/ | wc -l
0
```

Reply:
> Maybe add an unconditional panic!() so we have a test case?

> > Hmm, can you explain more what you mean?

I don't know when this line of code will be executed. Can you change it to an unconditional panic and document the standard library so I can see a backtrace of how it happens? If it never happens that probably means the whole thing is unreachable and we don't have to worry about it in the first place.

Reply:
> Can you change it to an unconditional panic and document the standard library so I can see a backtrace of how it happens? If it never happens that probably means the whole thing is unreachable and we don't have to worry about it in the first place.

@jyn514 It turns out it is reachable! It panicked while documenting `core`, specifically this path:

```
DEBUG rustdoc::clean::utils resolve_type(Path { global: false, res: Def(TyParam, DefId(0:140 ~ core[8787]::f32::{impl#0}::to_int_unchecked::Int)), segments: [PathSegment { name: "Int", args: AngleBracketed { args: [], bindings: [] } }] },HirId { owner: DefId(0:139 ~ core[8787]::f32::{impl#0}::to_int_unchecked), local_id: 18 })
```

I think that path is the `Int` in the return type of [`f32::to_int_unchecked()`](https://doc.rust-lang.org/1.50.0/std/primitive.f32.html#method.to_int_unchecked).

Reply:
I *think* passing a depth of `0` should probably be fine since IIUC paths that resolve to type parameters shouldn't be able to have type parameters themselves...

Reply:
Ok, that makes sense to me. Could you add a comment about a) when this happens and b) why the depth of 0 is ok?